dttools/src/envreplace.c

```diff
@@ -0,0 +1,103 @@
+/*
+  Copyright (C) 2019- The University of Notre Dame This software is
+  distributed under the GNU General Public License.  See the file
+  COPYING for details.
+
+
+  Given a file, replace environment variable with actual values
+  into new file. If no output-file is specified the original will
+  be replaced.
+
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "debug.h"
+#include "stringtools.h"
+#include "xxmalloc.h"
+
+void show_help(const char *exe) {
+	fprintf(stderr, "Usage:\n%s input-file [output-file]\n", exe);
+}
+
+int main(int argc, char **argv) {
+
+	if(argc < 2 || argc > 3) {
+		show_help(argv[0]);
+		fatal("ARGC %d: %s", argc, argv[1]);
+		//exit(1);
+	}
+
+
+	const char *input = argv[1];
+	FILE *INPUT = fopen(input, "r");
+	if (INPUT == NULL) {
+		fatal("unable to open %s: %s", input, strerror(errno));
+	}
+
+	char *output = NULL;
+	if(argc > 2) {
+		output  = xxstrdup(argv[2]);
+	} else {
+		output  = string_format("%s.XXXXXX",input);
+		int output_fd = mkstemp(output);
+		if (output_fd == -1){
+			fatal("could not create `%s': %s", output, strerror(errno));
+		}
+		close(output_fd);
+	}
+
+	FILE *OUTPUT = fopen(output, "w");
+	if (OUTPUT == NULL) {
+		fatal("unable to open %s: %s", output, strerror(errno));
+	}
+
+	
+	char variable[1024];
+	int var_index = 0;
+	int valid_var = 0;
+
+	char c = fgetc(INPUT);
+	while (c != EOF)
+	{
+		if (c == '$'){
+			valid_var = 1;
+		} else if (valid_var && (
+				(c >= (char)65 && c <= (char)90) ||  // A-Z
```

Comment:
Is this intended to only allow upper case? More generally, I think it would be better to use `isalpha()` and friends rather than raw ASCII codes

Reply:
I will change this. I am open to either just upper or all alpha characters. Going by :

http://pubs.opengroup.org/onlinepubs/9699919799/

Environment variables for use in Shell consists of uppercase, underscore, and digit (when not first). I am open to other options.

Reply:
Interesting, I did not know POSIX only requires uppercase. I'm fine with leaving the character set as is, but still prefer `isupper()` and `isdigit()`

Reply:
I agree. I have the code changed. I am just doing a couple of other cleanup options.